File: juliaPartonShower/Shower.jl
include("ShowerHelpers.jl")
include("SplittingFunctions.jl")
include("Kinematics.jl")
include("Constants.jl")

using Random 


Random.seed!(12345) # Set the seed for easier testing

function generateEmissions(Q::Float64, Qcut::Float64, aSover::Float64, branchType::Int8, masses::Vector{Float64})

    R1 = rand()
    R2 = rand()
    R3 = rand()
    R4 = rand()
    
    ems = Emission(0, 1, 0, 0, true, true)

    # Check for the branch type and get the appropiate z and t emissions
    if branchType == 1
        tmin = 3.99 * (mu[0]^2 + Qcut)
        ems.t, ems.ContinueEvolution = tEmission(Q, R1, aSover, tmin, Intgg, masses, branchType)
        ems.z = zEmission(ems.t, R2, aSover, Intgg, InvInt_gg, masses, branchType)
    elseif branchType == 2
        t0 = mu[0]^2 + Qcut
        t1 = mu[1]^2 + Qcut
        tmin = 0.99 * (2 * sqrt(t1 * t0) + t1 + t0)
        ems.t, ems.ContinueEvolution = tEmission(Q, R1, aSover, tmin, Intqq, masses, branchType)
        ems.z = zEmission(ems.t, R2, aSover, Intqq, InvInt_qq, masses, branchType)
    elseif branchType == 3
        tmin = 3.99 * (mu[0]^2 + Qcut)
        ems.t, ems.ContinueEvolution = tEmission(Q, R1, aSover, tmin, Intgq, masses, branchType)
        ems.z = zEmission(ems.t, R2, aSover, Intgq, InvInt_gq, masses, branchType)
    end
    
    # If the evoltion has been terminated, return the emission struc
    if ems.ContinueEvolution == false
        ems.z = 1
        return ems
    end

    # Get the transverse momentum squared
    ems.pTsq = transversemmSquared(ems.t, ems.z, branchType, masses)

    # Apply the minimum transverse momentum cut
    if ems.pTsq < Qcut 
        ems.Generated = false
    end

    # Apply the overestimate of the splitting function cut according to the branch
    if branchType == 1
       if R3 > Pgg(ems.z) > Pgg_over(ems.z)
            ems.Generated = false
       end
    elseif branchType == 2
        if R3 > Pqq(ems.z) > Pqq_over(ems.z)
            ems.Generated = false
       end
    elseif branchType == 3
        if R3 > Pgq(ems.z) > Pgq_over(ems.z)
            ems.Generated = false
       end
    end

    # Apply the alphaS oversitmate cut
    if R4 > getAlphaS(ems.t, ems.z, Qcut) / aSover
        ems.Generated = false
    end

    ems.phi = (2 * rand() - 1) * pi

    if ems.Generated == false
        ems.z = 1
    end

    return ems

end



# Functionn to evolve a specific particle given its children
function evolveParticle(pa::Particle, pb::Particle, pc::Particle, Qcut::Float64, aSover::Float64)

    tcuttoff = 4 # The curtoff for the evolution variable

    #emission = Emission(pa.t, 1, 0, 0, true, true) # Base values for the emisisons

    Q::Float64 = sqrt(pa.t) * pa.z # Get the initial scale for this evolution
    masses = [0, 0] # Get the masses which are set to 0 for now

    # This loops until either an emission is accepted and generated or until the evolution is terminated
    while true
        # Condition if the evolution scale is below the cuttoff to break
        if Q < sqrt(tcuttoff * Qcut^2)
            pa.ContinueEvolution = false
            return
        end
        # Branch for quark emiting. Currently only q -> qg
        if abs(pa.id) < 6 && abs(pa.id) > 0
            branchType = 2
            pb.m = mufunc(pa.m, Qg(pa.m))
            pc.m = Qg(pa.m)
            emission = generateEmissions(Q, Qcut, aSover, branchType, masses)
            pb.id = pa.id
            pc.id = 21
        # Condition for gluon emittion which needs competition
        elseif abs(pa.id) == 21
            branchType = 1
            pb.m = Qg(pa.m)
            pc.m = Qg(pa.m)

            emission = generateEmissions(Q, Qcut, aSover, branchType, masses)

            pb.id = 21
            pc.id = 21

            if emission.continueEvolution == false
                emission = Emission(0, 0, 0, 0,  false, false)
            end
            # Test for the possibility of a g -> qqbar emission
            branchType  = 3
            for flavor in range(1, 5)
                emissionTemp = generateEmissions(Q, Qcut, aSover, branchType, masses)

                # Accept the emission if the generated t value is greater 
                if emissionTemp.continueEvolution == true && emissionTemp.t > emission.t
                    emission = emissionTemp
                    pb.m = mufunc(pa.m, Qg(pa.m))
                    pc.m = mufunc(pa.m, Qg(pa.m))
                    pb.id = flavor
                    pc.id = -flavor
                end
            end

        end
        # Rescale the evolution variable
        Q = sqrt(emission.t)
        # Condition to break out of the loop: 
        # either an emission has been generated or the evolution is terminated
        if emission.Generated == true || emission.ContinueEvolution == false
            break
        end

    end


    if emission.continueEvolution == false
        pa.continueEvolution = false
        pa.status = 1 # Set the parent particle to final state if there is not emission
        pa.virtuality = 0 # The virtuality is zero if the particle has no emission
        return
    end

    pb.t = emission.t
    pc.t = emission.t
    pb.z = emission.z
    pc.z = 1 - emission.z 

    pb.alpha = pa.alpha * pb.z
    pc.alpha = pa.alpha * pc.z

    pb.phi = emission.phi 
    pc.phi = emission.phi

    pT = sqrt(emisison.pTsq)
    vmsq = getVirtMsq(emission.t, emission.z) # Get the virtuality of the emitting particle

    pa.virtuality = vmsq

    pa.status = -1 # Set the emitting particle to initial state for emission

    pb.pT = pT
    pc.pT = pT 

    pb.status = 1
    pc.status = 1
    pb.continueEvolution = true
    pc.continueEvolution = true 

    return
end


function showerParticle(jet::Jet, particle::Particle, Qmin::Float64, aSover::Float64)
    # Append the progenitor to the jet's list of particles
    push!(jet.AllParticles, particle)

    i = 1

    while i < 26
        # Check if the current index is out of bounds for the list of particles in the jet
        # If so, we are done with showering this jet
        if i > length(jet.AllParticles) -1
            return
        end
        pa = jet.AllParticles[i]
        # Get the child particles' templates
        pb = Particle(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, [], 1, 1, 0, 0, true, [particle], [])
        pc = Particle(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, [], 1, 1, 0, 0, true, [particle], [])
        evolveParticle(pa, pb, pc, Qmin, aSover)
        # If the evolution is terminated, end this branch
        if pa.continueEvolution == false
            i += 1
            continue
        end
        # If not then set the current index to the b particle and append the c particle to the list
        jet.AllParticles[i] = pb
        push!(jet.AllParticles, Pc)
        # Append the child particles, b and c, to the list of children in the parent particle
        push!(particle.children, [pb, pc])

    end

end


function showerEvent(event, Qmin::Float64, aSover::Float64)

    newEvent= Event([], [])
    plist = []
    jets = Jet[]
    # Go through the list of particles in the event and get the ones that will be showered and append them to a list
    # This is needed since both particles are needed for the initial evolution scale
    for p in event.Jets
        # Check if the particle is an electron or positron to skip it
        if abs(p.id) == 11
            push!(newEvent.AllParticles, p)
        elseif abs(p.id) < 6 && abs(p.id) > 0 && p.status == 1 # Only shower final state particles
            push!(plist, p)
        end 
    end

    # Set the initial evolution scale for the two progenitors
    plist[1].t, plist[2].t = EvolutionScale(plist[1], plist[2])

    for (i, p) in enumerate(plist)
        jet = Jet([], p)

        showerParticle(jet, p, Qmin, aSover)
        reconSudakovBasis(p, plist[end - (i - 1)])
        rotateMomentaLab(p, jet.AllParticles)

        push!(jets, jet)
    end

    newEvent.AllParticles = globalMomCons(newEvent.AllParticles, jets)

    return newEvent
end

File: pythonPartonShower/Shower.py
import numpy as np
import random as rand
from Classes import * # Import the classes
from Constants import * # Import the constants
from SplittingFunctions import * # Import the splitting functions
from Kinematics import * # Import the kinematic functions
from alphaS import * # ALphaS coupling constant
from ShowerHelpers import *
import math
import sys
rand.seed(12345) # Set the seed to rule out randomness in testing



# Define the function to generate the emissions
def GenerateEmissions (Q, Qcut, aSover, tfac, branchType, mu):

    # Generate the three randomn numbers to determine whether or not to accept the emission.
    R1 = rand.random()
    R2 = rand.random()
    R3 = rand.random()
    R4 = rand.random()
    

    # Get an empty emission object to hold emission data.
    Ems= emissioninfo(0,1, 0, 0, 0, True, True)

    # Get the t emission value
    match branchType:
        case 1:
            Ems.t, Ems.ContinueEvolve = tEmission(Q, Qcut, R1, aSover, tfac, tGamma_gg, mu, branchType)         
        case 2:
            Ems.t, Ems.ContinueEvolve = tEmission(Q, Qcut, R1, aSover, tfac, tGamma_qq, mu, branchType)         
        case 3:         
            Ems.t, Ems.ContinueEvolve = tEmission(Q, Qcut, R1, aSover, tfac, tGamma_gq, mu, branchType)       
        case 4: 
            Ems.t, Ems.ContinueEvolve = tEmission(Q, Qcut, R1, aSover, tfac, tGamma_qg, mu, branchType)
        case _:
            raise Exception('Invalid Branching option.')
            
    # Determine if there was a generated t emission   
    if Ems.ContinueEvolve == False:   
        Ems.z = 1
        Ems.pTsq = 0
        Ems.Vmsq = 0
        return Ems
    
    
    # Get the z emissiom, tranverse momentum squared, and the virtual mass squared
    match branchType:
        case 1:        
            Ems.z = zEmission(Ems.t, Qcut, R2, aSover, tGamma_gg, inversetGamma_gg, mu, branchType)
        case 2:            
            Ems.z = zEmission(Ems.t, Qcut, R2, aSover, tGamma_qq, inversetGamma_qq, mu, branchType)
        case 3:  
            Ems.z = zEmission(Ems.t, Qcut, R2, aSover, tGamma_gq, inversetGamma_gq, mu, branchType)   
        case 4:       
            Ems.z = zEmission(Ems.t, Qcut, R2, aSover, tGamma_qg, inversetGamma_qg, mu, branchType)   
        case _:
            raise Exception('Invalid Branching option.')
    
    # Get the transverse momentum squared and virtual mass squared
    Ems.Ptsq = transversemmsq(Ems.t, Ems.z, branchType, mu)
    Ems.Vmsq = virtualmass(Ems.t, Ems.z)

    # Determine whether the transverse momentum is physical
    if Ems.Ptsq < pT2min:
        if debug: print('Invalid transverse momentum')
        Ems.Generated = False
    

    # Determine whether or no to accept the t value, and inturn accept the emission
    match branchType:
        case 1:   
            if R3 > Pgg(Ems.z) / Pgg_over(Ems.z):
                Ems.Generated = False
        case 2:
            if R3 > Pqq(Ems.z) / Pqq_over(Ems.z):
                Ems.Generated = False    
        case 3:
            if R3 > Pgq(Ems.z) / Pgq_over(Ems.z):
                Ems.Generated = False 
        case 4:   
            if R3 > Pqg(Ems.z) / Pqg_over(Ems.z):
                Ems.Generated = False       
        case _:
            raise Exception('Invalid Branching option.')
                
    # Compare the alphaS value and overestimate to a random number to veto according to it
    if R4 > GetalphaS(Ems.t, Ems.z, Qcut) / aSover:
        Ems.Generated = False 
        
    
    # If any of the tests are true, then there is no emission and return these values
    if Ems.Generated == False:
        Ems.z = 1
        Ems.Ptsq = 0
        Ems.Vmsq = 0  

    
    return Ems

# The function to evolve a certain particle
# This is the old one for testing
def Evolve(pa, pslist, Qc, aSover):
    
    # Set the branch type for now. This is meant so one can easily switch splitting functions for testing
    # Cases: 1: g -> gg, 2: q -> qg, 3: g -> qqbar, 4: q -> gq
    branch_type = 2
    
    
    fac_t = 3.999 # Minimum value for the cutoff to try emissions
    tscalcuttoff = 4 # ACtual cuttoff
    
    # The miminum evolution scale
    t_min = Qc**2
    
    # Set the initial emission data class values.
    # These will be written over later
    Emission = emissioninfo(pa.E**2, 1, 0, 0, 0, True, True)
    
    # List to hold the particles/emisisons
    ps = []
    
    # The magnitude of the momentum of the parent particle
    pmag = np.sqrt(pa.Px**2 + pa.Py**2 + pa.Pz**2)
    
    # Get the starting evolution scale
    Emission.t = EvolutionScale(pslist[0], pslist[1])[0]

    
    Pb = Particle(0, 0, 0, 0, 0, 0, 0, 0, mufunc(pa.m, Qg(pa.m)), 0, 0, False)
    Pc = Particle(0, 0, 0, 0, 0, 0, 0, 0, Qg(pa.m), 0, 0, False)
    #masses = [0, Qg(pa.m)]
    masses = [mufunc(pa.m, Qg(max(Pb.m, Pc.m))), Qg(pa.m)]
    # Evolve the particle until the emission is past the cuttoff or another condition is met.
    while np.sqrt(Emission.t) * Emission.z > np.sqrt( tscalcuttoff * t_min) :
        
        # Get the emission values.
        Emission = GenerateEmissions(np.sqrt(Emission.t) * Emission.z, np.sqrt(t_min), aSover, fac_t, branch_type, masses)

        # Terminate the evolution if a requirement is met.
        # This then stops this branch's evolution
        if Emission.ContinueEvolve == False:
            # Add the magnitude of the quark with respect to its origina direction
            ps.append(Particle(pa.typ, 1, np.sqrt(Emission.t), Emission.z, 0, 0, 0, pmag, 0, pmag, 0, False))
            pa.ContinueEvolution = False
            return ps
        
        # Return emssions if the current one is past the cutt offf
        if Emission.t < tscalcuttoff * t_min:
            
            if debug: print('Hit Cuttoff, stopping Evolution')
            ps.append(Particle(pa.typ, 1, np.sqrt(Emission.t), Emission.z, 0, 0, 0, pmag, 0, pmag, 0, False))
            pa.ContinueEvolution = False
            return ps
        
        
        # Append the emissions and then continue
        if Emission.z != 1:
            
            Pt = np.sqrt(Emission.Ptsq) 
            Emission.phi = (2*rand.random() - 1)*np.pi # Generated phi value
            Ei = np.sqrt(( 1- Emission.z)**2 * pmag**2 + Pt**2)
            
            # Depending on the branch, append the appropiate particle
            if branch_type == 2 or branch_type ==1:
                p = Particle(21, 1, np.sqrt(Emission.t), Emission.z, Pt, Pt * np.cos(Emission.phi), Pt * np.sin(Emission.phi), (1 -Emission.z) * pmag, Pc.m, Ei, Emission.phi, Emission.ContinueEvolve)
            elif branch_type == 3:
                p = Particle(-3, 1, np.sqrt(Emission.t), Emission.z, Pt, Pt * np.cos(Emission.phi), Pt * np.sin(Emission.phi), (1 -Emission.z) * pmag, Pc.m, Ei, Emission.phi, Emission.ContinueEvolve)

            # Resacale the magnitude of the momnetum with the z emission
            pmag = Emission.z * pmag
            ps.append(p)
    
    # Add the magnitude of the quark with respect to its origina direction.
    ps.append(Particle(pa.typ, 1, np.sqrt(Emission.t), Emission.z, 0, 0, 0, pmag, 0, pmag, 0, False))
    return ps

# New function to evolve a particle with the competition.
# I figured it would be better to create a seperate function so I can switch back to the old one for testing.
# The function to evolve a certain particle
def EvolveParticle(Pa, Pb, Pc, Qc, aSover):
    tMin = Qc**2
    fac_t = 3.999 # Minimum value for the cutoff to try emissions
    tscalcuttoff = 4# ACtual cuttoff
    # Set the initial emission data class values
    Emission = emissioninfo(Pa.t_at_em, 1, 0, 0, 0, True, True)
    
    # Set the evolution variable from the parent particle's information
    Q = np.sqrt(Pa.t_at_em) * Pa.z_at_em
    masses = [0, 0] # Set the masses to 0 for testing
    # Loop until either the particle evolution is terminated or an emission is generated
    # This is an implementation of a do while loop from another codebase 
    while True:
        # If the evolution variables is below the cutoff, 
        if Q < np.sqrt(tscalcuttoff * Qc**2):
            Pa.ContinueEvolution = False
            return
        
        # Code to determine which branch is in effect. 
        # If the parent particle is a quark, then the outgoing quark is the same type
        # For now there is only q -> qg for initial quark
        if abs(Pa.typ) < 6 and abs(Pa.typ) > 0:
            # Set the branch type and get the masses to generate the emissions.
            branch_type = 2
            Pb.m = mufunc(Pa.m, Qg(Pa.m))
            Pc.m = Qg(Pa.m)
            #masses = [Pb.m, Qg(Pa.m)]
            Emission = GenerateEmissions(Q, np.sqrt(tMin), aSover, fac_t, branch_type, masses)
            Pb.typ = Pa.typ
            Pc.typ = 21
            
            
        # For gluons, there are g -> gg and g -> qqbar.
        # So a competition for which to accept is needed.
        elif abs(Pa.typ) == 21:
            # Set the branch type and get the masses to generate the emissions.
            branch_type = 1
            Pb.m = Qg(Pa.m)
            Pc.m = Qg(Pa.m)
            #masses = [Pb.m, Qg(Pa.m)]

            # Get emission for g -> gg
            Emission = GenerateEmissions(Q, np.sqrt(tMin), aSover, fac_t, branch_type, masses)
            
            # Set the child particles type
            Pb.typ = 21
            Pc.typ = 21

            
            
            # If stopped evolution, set t to 0 and continue for test of g -> qqbar emision
            if Emission.ContinueEvolve == False:
                
                Emission.t = 0
            
            # Go through each of the quark flavors and test whether one is emitted and the t emission is greater than the g -> gg t emission
            for flavor in range(1, 6):
                # Get emission for g -> qqbar
                branch_type = 3
   
                #masses = [mufunc(Pa.m, Qg(Pa.m)), Qg(Pa.m)]
                EmissionTemp = GenerateEmissions(Q, np.sqrt(tMin), aSover, fac_t, branch_type, masses)
                
                # Accept g -> qqbar emission if true.
                if EmissionTemp.ContinueEvolve == True and EmissionTemp.t > Emission.t:
                    Pb.m = mufunc(Pa.m, Qg(Pa.m))
                    Pc.m = mufunc(Pa.m, Qg(Pa.m))
                    Pb.typ = flavor
                    Pc.typ = -flavor
                    Emission = EmissionTemp
                    
                    
        else:
            print('ERROR: Invalid Particle type')

        Q = np.sqrt(Emission.t)
        if Emission.Generated == True or Emission.ContinueEvolve == False:
            break
    # Terminate the evolution if needed.
    # This then stops this branch's evolution
    if Emission.ContinueEvolve == False:
        Pa.ContinueEvolution = False
        return
    # Generated the phi value for this emission
    Emission.phi = (2*rand.random() - 1)*np.pi 
    
    # Set the children particles' t and z emission values
    Pb.t_at_em = (Emission.t)
    Pc.t_at_em = (Emission.t)
    Pb.z_at_em = Emission.z
    Pc.z_at_em = 1- Emission.z
    
    # Get transverse momentum and momentum magnitude of mother particle
    Pt = np.sqrt(Emission.Ptsq)
    pmag = np.sqrt(Pa.Px**2 + Pa.Py**2 + Pa.Pz**2)
    
    # Set the b child particle's momentum and energy values.
    Pb.Px = Pt * np.cos(Emission.phi)
    Pb.Py = Pt * np.sin(Emission.phi)
    Pb.Pz = Emission.z* pmag
    Pb.E = np.sqrt(Pb.Px**2 + Pb.Py**2 + Pb.Pz**2)
    
    # Set the c child particle's momentum and energy values.
    Pc.Px = -Pt * np.cos(Emission.phi)
    Pc.Py = -Pt * np.sin(Emission.phi)
    Pc.Pz = (1- Emission.z) * pmag
    Pc.E = np.sqrt(Pc.Px**2 + Pc.Py**2 + Pc.Pz**2)
    #print(Emission.t)
    # Set the other variables for the particles
    Pb.status = 1
    Pc.status = 1
    Pb.Pt = Pt
    Pc.Pt = Pt
    Pb.ContinueEvolution = True
    Pc.ContinueEvolution = True

    
    # Rotate the child particles to allign with the parent particle
    RotatedParticles = RotateMomentaLab(Pa, [Pb, Pc])
    Pb = RotatedParticles[0]
    Pc = RotatedParticles[1]
    
    return 

def ShowerParticle(jet, particle, Qmin, aSover):
    # Append the parent particle to jet's particles
    jet.Particles.append(particle)
    
    i =0
    
    
    while i < 25:
        # Check if the index is out of bounds for the list
        if i > len(jet.Particles) - 1:
            return
        Pa = jet.Particles[i]
        # Get children particles' template
        Pb = Particle(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, False)
        Pc = Particle(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, False)
        EvolveParticle(Pa, Pb, Pc, Qmin, aSover)
        
        if Pa.ContinueEvolution == False:
            i += 1
            continue
        jet.Particles[i] = Pb
        jet.Particles.append(Pc)
    
    # Recursive soltion. Doesn't work for some reason
    '''
    Pa = (particle)
    # Get children particles' template
    Pb = Particle(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, False)
    Pc = Particle(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, False)
    # Evolve the particles
    EvolveParticle(Pa, Pb, Pc, Qmin, aSover)
    # Escape if evolution is terminated
    if Pa.ContinueEvolution == False:
        return
    else:
    # Else append the children particles to the jet and shower them recursively
        ShowerParticle(jet, Pb, Qmin, aSover)
        ShowerParticle(jet, Pc, Qmin, aSover)
    '''
    return

def ShowerEvent(event, Qmin, aSover):
    
    NewEvent = Event([],[])
    plist = []
    
    # Go through the list of progenitors in the event and shower each
    for p in event.Jets:

            # Test if particle is electron or positron and then append it
            if abs(p.typ) == 11:
                NewEvent.AllParticles.append(p)
                continue
            elif abs(p.typ) < 6 and abs(p.typ) > 0 and p.status==1:
                plist.append(p)
    a = EvolutionScale(plist[0], plist[1])
    plist[0].t_at_em = a[0]
    plist[1].t_at_em =a[1]
 
    
    for p in plist:
            jet = Jet(p, [])

            # Shower the progenitor
            ShowerParticle(jet, p, Qmin, aSover)   
            
            NewEvent.Jets.append(jet)

    
    # Apply global momentum for this event
    NewEvent.AllParticles = Glob_mom_cons(NewEvent.AllParticles, NewEvent.Jets)
    
    return NewEvent

# This is for testing as this does not involve the new Evolution function with competition
# Define the testing function to shower the events
def Shower_Evens(Event, Qmin, aSover):
    
    AllParticles = []
    Jets = []
    pslist = []
    # Go through the first event and shower each particle
    # This was for testing/comparison
    for i in Event.Jets:
        # Test to only evolve quarks
        if abs(i.typ) == 11:
            AllParticles.append(i)
            continue
        elif abs(i.typ) < 6 and abs(i.typ) > 0 and i.status==1:
            pslist.append(i)
            
    for i in pslist:
            ps = Evolve(i, pslist, Qc, aSover )
            
            # Rotate the particle with the lab
            rotated = RotateMomentaLab(i, ps)
            AllParticles.extend(rotated)
            #AllParticles.append(rotated)
            jet = Jet(i, rotated)
            Jets.append(jet)
    AllParticles = Glob_mom_cons(AllParticles, Jets)

    return AllParticles, Jets

